pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -1729,7 +1729,10 @@ private void getLargestBatchIndexWhenPossible(
         ManagedLedgerImpl ml = (ManagedLedgerImpl) persistentTopic.getManagedLedger();
 
         // If it's not pointing to a valid entry, respond messageId of the current position.
-        if (lastPosition.getEntryId() == -1) {
+        // If the compaction cursor reach the end of the topic, respond messageId from compacted ledger
+        Optional<Position> compactionHorizon = persistentTopic.getCompactedTopic().getCompactionHorizon();
+        if (lastPosition.getEntryId() == -1 || (compactionHorizon.isPresent()
+                        && lastPosition.compareTo((PositionImpl) compactionHorizon.get()) <= 0)) {
```

Comment:
Should it be `lastPosition.compareTo((PositionImpl) compactionHorizon.get()) >= 0` ?
Does the unit test cover this?

Reply:
- compaction horizon < last message -> the compaction cursor does not reach the end of the original topic, which means not all the data has been compacted, so we should return the last message ID from the original topic.
- compaction horizon >= last message -> the compaction cursor reach the end of the original topic, which means all the data been has been compacted, so we should return the last message ID from the compacted ledger.

Reply:
Yes, the unit tests can cover the horizon == last message ID, theoretically, horizon should not > last message ID